src/streaming/zarr.common.cpp

```diff
@@ -0,0 +1,297 @@
+#include "macros.hh"
+#include "zarr.common.hh"
+
+#include <stdexcept>
+
+size_t
+zarr::bytes_of_type(ZarrDataType data_type)
+{
+    switch (data_type) {
+        case ZarrDataType_int8:
+        case ZarrDataType_uint8:
+            return 1;
+        case ZarrDataType_int16:
+        case ZarrDataType_uint16:
+            return 2;
+        case ZarrDataType_int32:
+        case ZarrDataType_uint32:
+        case ZarrDataType_float32:
+            return 4;
+        case ZarrDataType_int64:
+        case ZarrDataType_uint64:
+        case ZarrDataType_float64:
+            return 8;
+        default:
+            throw std::runtime_error("Invalid data type: " +
+                                     std::to_string(data_type));
+    }
+}
+
+size_t
+zarr::bytes_of_frame(const std::vector<Dimension>& dims, ZarrDataType type)
+{
+    return bytes_of_type(type) * dims.back().array_size_px *
+           dims[dims.size() - 2].array_size_px;
+}
+
+size_t
+zarr::chunks_along_dimension(const Dimension& dimension)
+{
+    EXPECT(dimension.chunk_size_px > 0, "Invalid chunk size.");
+
+    return (dimension.array_size_px + dimension.chunk_size_px - 1) /
+           dimension.chunk_size_px;
+}
+
+size_t
+zarr::shards_along_dimension(const Dimension& dimension)
+{
+    const size_t shard_size = dimension.shard_size_chunks;
+    if (shard_size == 0) {
+        return 0;
+    }
```

Comment:
Nit: Add a check for zero value before the assignment to prevent an unnecessary copy.
```cpp
if (dimension.shard_size_chunks == 0) {
    return 0;
}
const auto shard_size = dimension.shard_size_chunks;
```